kernel/src/fs/cgroupfs/controller/mod.rs

```diff
@@ -0,0 +1,413 @@
+            return false;
+        };
+
+        let guard = disable_preempt();
+        let Some(sub_controller) = self.obtain_sub_controller_with(ctrl_type, &guard) else {
+            // If the sub-controller is not active, all its attributes are considered absent.
+            return true;
+        };
+
+        sub_controller.attr_set().get(name).is_none()
+    }
+
+    pub(super) fn read_attr_at(
+        &self,
+        name: &str,
+        offset: usize,
+        writer: &mut VmWriter,
+        cgroup_node: &dyn CgroupSysNode,
+    ) -> Result<usize> {
+        let Some((subsys, _)) = name.split_once('.') else {
+            return Err(Error::NotFound);
+        };
+        let ctrl_type = ControllerType::try_from(subsys)?;
+
+        let guard = disable_preempt();
+        let Some(sub_controller) = self.obtain_sub_controller_with(ctrl_type, &guard) else {
+            return Err(Error::NotFound);
+        };
+
+        sub_controller.read_attr_at(name, offset, writer, cgroup_node)
+    }
+}
+
+/// A locked controller for a cgroup.
+///
+/// Holding this lock indicates exclusive access to modify the sub-control state.
+pub(super) struct LockedController<'a> {
+    sub_ctrl_state: MutexGuard<'a, SubCtrlState>,
+    controller: &'a Controller,
+}
+
+impl LockedController<'_> {
+    /// Activates a sub-control with given name.
+    pub(super) fn activate(
+        &mut self,
+        ctrl_type: ControllerType,
+        current_node: &dyn CgroupSysNode,
+        parent_controller: Option<&LockedController>,
+    ) -> Result<()> {
+        if self.sub_ctrl_state.is_active(ctrl_type) {
+            return Ok(());
+        }
+
+        // A cgroup can activate the sub-control only if this
+        // sub-control has been activated in its parent cgroup.
+        if parent_controller
+            .is_some_and(|controller| !controller.sub_ctrl_state.is_active(ctrl_type))
+        {
+            return Err(Error::NotFound);
+        }
+
+        self.sub_ctrl_state.activate(ctrl_type);
+        self.updata_sub_controller(ctrl_type, current_node);
+
+        Ok(())
+    }
+
+    /// Deactivates a sub-control with given name.
+    pub(super) fn deactivate(
+        &mut self,
+        ctrl_type: ControllerType,
+        current_node: &dyn CgroupSysNode,
+    ) -> Result<()> {
+        if !self.sub_ctrl_state.is_active(ctrl_type) {
+            return Ok(());
+        }
+
+        // If any child node has activated this sub-control,
+        // the deactivation operation will be rejected.
+        let mut can_deactivate = true;
+        // This is race-free because if a child wants to activate a sub-controller, it should first
+        // acquire the lock of the parent controller, which is held here.
+        current_node.visit_children_with(0, &mut |child| {
+            let cgroup_child = child.as_any().downcast_ref::<CgroupNode>().unwrap();
+            let child_controller = cgroup_child.controller().lock();
```

Comment:
This is a mutex. It cannot be locked after locking a spin lock (`visit_children_with`).